File: SudokuSolver/SudokuSolver/Backtrack.cs
using System.Collections.Generic;

namespace SudokuSolver
{
    class Backtrack
    {
        private Stack<Sudoku> fStack;
        private int fCheckCount;

        public Backtrack(Sudoku lInitial)
        {
            fStack = new Stack<Sudoku>();
            fStack.Push(lInitial);
            fCheckCount = 0;
        }

        public int CheckCount
        {
            get { return fCheckCount; }
        }

        public Sudoku Solve()
        {
            while (fStack.Count > 0)
            {
                Sudoku lSudoku = fStack.Pop();

                if (lSudoku.Solved())
                {
                    return lSudoku;
                }

                int lBlankIndex;

                for (lBlankIndex = 0; lBlankIndex < 81; lBlankIndex++)
                {
                    if (lSudoku.ValueAt(lBlankIndex) == 0)
                    {
                        break;
                    }
                }

                for (int i = 9; i > 0; i--)
                {
                    if (lSudoku.CanSetValue(lBlankIndex, i))
                    {
                        Sudoku lMoveSudoku = new Sudoku(lSudoku);
                        lMoveSudoku.SetValue(lBlankIndex, i);
                        fStack.Push(lMoveSudoku);
                    }
                }

                fCheckCount++;
            }

            return null;
        }
    }
}

File: archive/SudokuSolver/SudokuSolverCPP/Backtrack.cpp
#include "Backtrack.h"

Backtrack::Backtrack(Sudoku aSudoku)
{
	fStack.push(aSudoku);
	fCheckCount = 0;
}

std::string Backtrack::Solve(void)
{
	while (fStack.size() > 0)
	{
		Sudoku lSudoku = fStack.top();
		fStack.pop();

		if (lSudoku.Solved())
		{
			return lSudoku.ToString();
		}

		int lBlankIndex;

		for (lBlankIndex = 0; lBlankIndex < (SIZE * SIZE); lBlankIndex++)
		{
			if (lSudoku.ValueAt(lBlankIndex) == 0)
			{
				break;
			}
		}

		for (int i = SIZE; i > 0; i--)
		{
			if (lSudoku.CanSetValue(lBlankIndex, i))
			{
				Sudoku lMoveSudoku(lSudoku);
				lMoveSudoku.SetValue(lBlankIndex, i);
				fStack.push(lMoveSudoku);
			}
		}

		fCheckCount++;
	}

	return "No solution found...";
}

int Backtrack::GetCheckCount(void)
{
	return fCheckCount;
}
